repair/row_level.cc

```diff
@@ -302,13 +303,20 @@ mutation_reader repair_reader::make_reader(
             // So we release the one on _permit so the only one is the one the
             // shard reader will obtain.
             _permit.release_base_resources();
-            return make_multishard_streaming_reader(db, _schema, _permit, [this] {
+            auto rd = make_multishard_streaming_reader(db, _schema, _permit, [this] {
                 auto shard_range = _sharder.next();
                 if (shard_range) {
                     return std::optional<dht::partition_range>(dht::to_partition_range(*shard_range));
                 }
                 return std::optional<dht::partition_range>();
             }, compaction_time);
+            // Setting the repair buffer size as the multishard reader's buffer
+            // size helps avoid extra cross-shard round-trips and possible
+            // evict-recreate cycles.
+            // Cap the reader's buffer size at 128K to avoid too much memory
+            // consumption during the read.
+            rd.set_max_buffer_size(std::min(max_buffer_size, size_t(128*1024)));
```

Comment:
This is effectively always setting the max buffer size to 128K. The repair buffer which contains multiple repair_rows  is much larger than 128K, it is 32MiB.  If 128K for reader buffer is good, it looks so according to the test, that's great. What is the buffer size will the reader user before this PR?

Reply:
8K

I want to use as large a buffer as I can, without causing trouble in other places. I have to investigate the failed tests to see why they fail after increasing the buffer size.

Reply:
OK.  A 16X bigger reader buffer helps to make the reader more efficient. Could you please explain in more detail how and why it helped exactly? Why does it help to reduce the cross shard round trips and reader eviction and recreation.

Reply:
Isn't it self-explanatory why reducing number of cross-shard round-trips helps? Cross shard trip has an overhead, you make less trips -> less overhead.
Recreating the readers is another expensive operation, the less you have of this, the better.

Reply:
I am asking why increasing the buffer size reduce the cross shard round trips and reduce the recreation?

Reply:
From the PR description:
> Add a buffer hint to the multishard reader. This is an internal hint, used by the multishard reader to provide a hint to the shard reader, on how much data exactly is needed by the multishard reader from the respective shard. This hint allows eliminating extraneous cross-shard round-trips and possible shard reader evict-recreate cycles. Building on this, repair sets its own row buffer size as the max buffer size on the multishard reader, ensuring that the row buffer is filled with the minimum amount of cross-shard round trips and minimal reader recreation.

The multishard reader now uses the buffer size and propagates to the shard reader to read exact amount of data that was requested, to complete the read in a single shard-trip.

Reply:
Yes, now the shard reader reads the exact amount the multi shard reader asks it to read. But how does this have impact on the cross shard round trips? What are the cross shard ops exactly? Just copying buffer read from a remote shard to shard where the multi shard reader is.  What amount will the shard reader before this patch? Why it is wasteful before.  If the buffer size is bigger, doesn't it drop more data and waste more when a reader is evicted.

Reply:
Before this PR, each reader in the tree had a buffer of size 8K. So if repair wanted to fill its 32MB buffer, it had to call `fill_buffer()` on the reader 32MB/8K=4K times (the `fill_buffer()` call is hidden, alled by `operator()`). Setting the multishard reader's buffer size to something larger would have not helped: shard readers would continue to read 8K per buffer, which needs doing the following 4K times: go to shard N, recreate reader if it was evicted, call `fill_buffer()`, copy results back.
After the patch, multishard reader calculates how much is missing from current buffer, then pass this size to shard buffer, so there will be exactly one of the round-trip described above.
So after the patch, setting the multishard reader's buffer size to the same as that of the repair row buffer, will reduce number of shard round trips by 4K. If we set to 128K, it reduces it by 256 (still nice reduction).

Reply:
Wow. I wonder how our recently introduced RPC statistics could have helped here.

Reply:
Setting to 128K, it reduces by 16, not 256.
```
Before: 32M / 8K =4K
After: 32M / 128K=256 
Before/After: 4K / 256 = 16
```

The number of total bytes of remote shard buffer copy does not change, using 8K or 128K shard reader buffer, since the total amount does not change. I think the real dominating factor is that we have less number of shard reader recreation. I think even the chance of the shard reader eviction does not change much (yes the total number of eviction changes).  Roughly, total eviction = total_reader_read  * eviction_probability.

Reply:
> Wow. I wonder how our recently introduced RPC statistics could have helped here.

This kind of cross-shard back-and-forth will not show up there, this happens internally on a single node.